drivers/fpga/fpga_mpfs.c

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2024 Microchip Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_mpfs_fpga
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/fpga.h>
+#include <zephyr/kernel.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(fpga_mpfs);
+
+#define SPI_FLASH_DIRECTORY_OFFSET    0x00000000
+#define SPI_FLASH_GOLDEN_IMAGE_OFFSET 0x00100400
+#define SPI_FLASH_NEW_IMAGE_OFFSET    0x01500400
+#define SPI_FLASH_SECTOR_SIZE         4096
+#define SPI_FLASH_PAGE_SIZE           256
+
+#define MSS_SCBMAILBOX      ((volatile uint32_t *)(0x37020800UL))
+#define MSS_SCB_SERVICES_CR (*(volatile uint32_t *)(0x37020050UL))
+#define MSS_SCB_SERVICES_SR (*(volatile uint32_t *)(0x37020054UL))
```

Comment:
IMO we should be reading these addresses from the device tree. Maybe a syscon node/driver would be a good fit here?

Reply:
+1

Reply:
I agree. We will look at this, thanks!

Reply:
@carlocaione @fkokosinski 
Thanks for this one guys, we have went with a different approach